### PR TITLE
Refactor await and yield implementation and enable async iteration

### PR DESCRIPTION
**Pull request by zenparsing**
## Overview

The goals of this change are to:

- Enable async generators
- Streamline and simplify the relationship between generators, async functions, and async generators.
- Make async functions and generators more JIT-friendly by reducing the number and complexity of supporting helpers.
- Improve code readability and maintainability.

The primary architectural change is to perform yield resume logic and branching directly in bytecode, rather than relying on the `ResumeYield` helper, and to desugar all async functions and async generators to normal generators. This allows us to eliminate the special-case `GeneratorReturnException` exception type.

In order to perform yield resume logic in bytecode, we resume the generator by creating a dynamic object with "value" (an arbitrary var) and "kind" (tagged int) fields. Branch instructions are inserted to deal with the various "kind" cases: "normal", "return", and "throw".

To desugar async functions and async generators to normal generators, we introduce a new type ID to represent "await results". Async generators then test for this type ID when determining whether their internal coroutine is "awaiting" or "yielding".

### Architectural

- Remove `ResumeYieldData`. Instead, resume the generator with a "resume yield object" containing "value" and "kind" properties.
- Remove `GeneratorReturnException`. The bytecode is now responsible for returning based on the resume yield kind.
- Remove opcodes `ResumeYield`, `Await`, `AsyncYield`, `AsyncYieldStar`, `AsyncYieldIsReturn`.
- Add new opcode and helper for `NewAwaitObject`. This object type is yielded when awaiting and is uniquely distinguished by type ID from "normal" yield results.
- Remove `ResumeYield` and `ResumeYieldStar` opcodes and helpers. When generators are resumed, the resume yield object is loaded into the yield register and the resume logic is performed in bytecode.
- Make async generators a subclass of `JavascriptGenerator`.
- Enable flag `ES2018AsyncIteration` by default.

### JIT

- Remove unused `GeneratorResumeYieldLabel`.
- Rename `GeneratorLoadResumeYieldData` to `GeneratorResumeYield`.
- Remove support for `ResumeYield`. `GeneratorResumeYield` is inserted immediately after the bail-in instr and takes care of loading the resume yield object.

### ByteCode

- Rename `EmitDummyYield` to `EmitStartupYield`.
- Remove unnecessary wrapping of async function body in try/catch. Exception handling is performed by library code that wraps the coroutine.
- Move startup yield after function declaration hoisting to unify module and generator code paths.
- Implement iterator next method caching to match spec.
- Replace `tryScopeRecordsList` and loop "parent node linking" with a single mechanism for performing cleanup in both the branch-in-loop cases and yield cases.
- Implement yield star evaluation logic in bytecode.
- Implement yield resume branching in bytecode.
- Await `return` operand in async generators to match spec.
- Add new error message for when `yield *` inner iterator does not have a "throw" method.

### Parser

- Remove `ParseNodeLoop`. The only purpose of this node type was to store an outer statement reference used when emitting jump cleanup. This information is now maintained in the bytecode generator's `jumpCleanupList`.

### Library

- Move `JavascriptAsyncGeneratorFunction` into a new file.
- Add type ID for async generators.
- Remove unnecessary `JavascriptPromiseAsyncSpawnExecutorFunction`.
- Use more readable names for `CreateIteratorResultObject` functions.
- Use more readable names for async function support classes.
- Move `JavascriptAsyncFunction` into a new file (previously code was split between `JavascriptPromise` and `JavascriptGeneratorFunction`).
- Move `JavascriptAsyncGeneratorFunction` into a new file.
- Create locked library types for await objects and resume yield objects.
- Move "copy arguments to heap" logic into `JavascriptGenerator::New`.

## Performance

This change results in a small performance increase on the ARES-6 benchmark.

| Scenario | Average Time | Change over baseline |
|----------|--------------|----------------------|
| ARES-6/Basic (No JIT) |	3671ms | 3.96% faster |
| ARES-6/Basic (JIT) | 2502ms | 2.56% faster |

### CODE REVIEW

**Comment by rhuanjl** [comment]
It's nice to see the horrible OP_ResumeYield logic replaced.

A little sad to see that much of my implementation of AsyncGenerators is removed/re-written here BUT I can see that this way of doing it is a lot neater.

A couple of things stand out to me by way of future improvements:
 - The startup yield is only needed when the generator has complex parameters - there was code to put a relevant condition around it in the PR to enable jitting generators - obviously that PR as it stands is dead with this change but that enhancement seems worth bringing across to me.
 - In my asyncGenerator implementation I cached the continuation methods for the async generator rather than creating them every time obviously doing that has some memory cost but I figure it should be worth it for performance and the same could/should be done for async functions.
 - Similarly the await object should only need to be created once and could be cached in an internal slot
 - The caching of the iterator Next property is missing in some other places not just generators

**Comment by zenparsing** [comment]
Hi @rhuanjl - I was a bit nervous about posting this PR given there were so many changes to what you had done, which was awesome work given the existing generators implementation. But particularly with JITing coming I want to take this opportunity to clean up the existing oddness. Apologies!

> In my asyncGenerator implementation I cached the continuation methods for the async generator rather than creating them every time obviously doing that has some memory cost but I figure it should be worth it for performance and the same could/should be done for async functions.

Oh right, I had forgotten about that, and it was a good idea. Hopefully this change will make it a bit easier to apply the same optimization to async functions (since there are fewer continuation functions to worry about).

> Similarly the await object should only need to be created once and could be cached in an internal slot

Yes! I was trying to figure out whether it would be better to store it in a C++ field or create a register for it in bytecode, but I didn't try to implement it either way yet.

> The caching of the iterator Next property is missing in some other places not just generators

Right. My idea was to go ahead and take care of the bytecode part here, and then use your existing PR for the remaining changes to the library code. I *think* I got all of the bytecode here?

**Comment by zenparsing** [comment]
Those changes look great! We can bring them in after this one gets landed.

For the TTD stuff, I don't really know it either but I think @boingoing could help when we get there. I'm also trying to consider what we want to do with TTD longer-term.

**Comment by nhat-nguyen** [comment]
Awesome!! 😊😊

**Comment by fatcerberus** [comment]
> Remove unnecessary wrapping of async function body in try/catch. Exception handling is performed by library code that wraps the coroutine.

Ooh! Will that fix #4630, I wonder...? 🤞

**Comment by rhuanjl** [comment]
> > Remove unnecessary wrapping of async function body in try/catch. Exception handling is performed by library code that wraps the coroutine.
> 
> Ooh! Will that fix #4630, I wonder...? 🤞

It doesn't - BUT it lays the groundwork for an easy fix of #4630 that I have ready to submit after this lands (if someone doesn't beat me to it).

File: lib/Runtime/Library/JavascriptGenerator.cpp [hunk 1]
```diff
@@ -5,857 +5,509 @@
...
-            return promise;
-        }
+void JavascriptGenerator::ExtractSnapObjectDataInto(TTD::NSSnapObjects::SnapObject* objData, TTD::SlabAllocator& alloc)
+{
+    TTD::NSSnapObjects::SnapGeneratorInfo* gi = alloc.SlabAllocateStruct<TTD::NSSnapObjects::SnapGeneratorInfo>();
 
-        // 4. Let queue be generator.[[AsyncGeneratorQueue]].
-        // 5. Let request be AsyncGeneratorRequest { [[Completion]]: completion, [[Capability]]: promiseCapability }.
-        AsyncGeneratorRequest* request = RecyclerNew(scriptContext->GetRecycler(), AsyncGeneratorRequest, input, exceptionObj, promise);
+    // TODO: BUGBUG - figure out how to determine what the prototype was
```
Comment (boingoing) [inline lib/Runtime/Library/JavascriptGenerator.cpp]: Is this ok? Is TTD broken with this?

Comment (nhat-nguyen) [reply lib/Runtime/Library/JavascriptGenerator.cpp]: @boingoing I believe TDD has always been broken on this path. I remember seeing this same comment but didn't have enough context to understand it further.

File: lib/Parser/Parse.cpp [hunk 2]
```diff
@@ -7322,7 +7322,6 @@ void Parser::ParseExpressionLambdaBody(ParseNodeFnc * pnodeLambda, bool fAllowIn
         // Pushing a statement node with PushStmt<>() normally does this initialization
         // but do it here manually since we know there is no outer statement node.
         pnodeRet->grfnop = 0;
-        pnodeRet->pnodeOuter = nullptr;
```
[the remark this answers is not in the document]
Comment (zenparsing) [reply lib/Parser/Parse.cpp]: It was only used for the `EmitJumpCleanup` routine in the byte code generator. After joining the "jump cleanup" paths with "generator yield cleanup" paths, we don't need this state in the parse tree anymore.

File: lib/Runtime/ByteCode/ByteCodeEmitter.cpp [hunk 3]
```diff
@@ -6211,11 +6147,121 @@ void EmitReference(ParseNode *pnode, ByteCodeGenerator *byteCodeGenerator, FuncI
...
 }
 
-void EmitGetIterator(Js::RegSlot iteratorLocation, Js::RegSlot iterableLocation, ByteCodeGenerator* byteCodeGenerator, FuncInfo* funcInfo, bool isAsync = false);
-void EmitIteratorNext(Js::RegSlot itemLocation, Js::RegSlot iteratorLocation, Js::RegSlot nextInputLocation, ByteCodeGenerator* byteCodeGenerator, FuncInfo* funcInfo);
+void EmitGetIterator(Js::RegSlot iteratorReg, Js::RegSlot iterableReg, ByteCodeGenerator* byteCodeGenerator, FuncInfo* funcInfo);
 void EmitIteratorClose(Js::RegSlot iteratorLocation, ByteCodeGenerator* byteCodeGenerator, FuncInfo* funcInfo, bool isAsync = false);
-void EmitIteratorComplete(Js::RegSlot doneLocation, Js::RegSlot iteratorResultLocation, ByteCodeGenerator* byteCodeGenerator, FuncInfo* funcInfo);
-void EmitIteratorValue(Js::RegSlot valueLocation, Js::RegSlot iteratorResultLocation, ByteCodeGenerator* byteCodeGenerator, FuncInfo* funcInfo);
+
+void EmitFunctionCall(
```
[the remark this answers is not in the document]
Comment (zenparsing) [reply lib/Runtime/ByteCode/ByteCodeEmitter.cpp]: There was quite a bit of code movement here in order to better support calling iterator methods from bytecode. For example, we needed to cache the "next" method of iterators and use that method on subsequent calls in for-of and yield star. To that end, `EmitIteratorNext` and `EmitIteratorValue` were removed in favor of the more general `EmitFunctionCall` and `EmitGetObjectProperty`.

File: lib/Runtime/ByteCode/ByteCodeGenerator.cpp [hunk 4]
```diff
@@ -4724,6 +4726,21 @@ void CheckFuncAssignment(Symbol * sym, ParseNode * pnode2, ByteCodeGenerator * b
     };
 }
 
+void AssignYieldResumeRegisters(ByteCodeGenerator* byteCodeGenerator)
+{
+    // On resuming from a yield, we branch based on the ResumeYieldKind
+    // integer value
+    byteCodeGenerator->EnregisterConstant((uint)Js::ResumeYieldKind::Normal);
```
[the remark this answers is not in the document]
Comment (zenparsing) [reply lib/Runtime/ByteCode/ByteCodeGenerator.cpp]: Thanks for asking about this.

These constants are used to determine how the iterator was resumed. If the generator was resumed with a "normal completion", then the resume object will have a kind property whose value is `(uint)Js::ResumeYieldKind::Normal`. (Similarly for "throw" and "return".) Bytecode is generated that checks the "kind" property against these constant values.

Do you think this is a good approach?

File: lib/Runtime/Library/JavascriptGenerator.cpp (hunk 1 above)
Comment (zenparsing) [reply lib/Runtime/Library/JavascriptGenerator.cpp]: Yes. All of this, including the TTD stuff (which I've got my eye on for possible simplification) was just copied over.

**Comment by pleath** [comment]
Note that the *_ReuseLoc variants should not be used the first time a code sequence writes to a register, only the 2nd and subsequent times. So:
R3 = LdFld R4.x
R3 = LdFld_ReuseLoc R5.x
Not:
R3 = LdFld_ReuseLoc R4.x
R3 = LdFld_ReuseLoc R5.x

The reason is that _ReuseLoc tells the JIT not to start a new lifetime. So in the above example, each _ReuseLoc op extends an existing lifetime of R3.

**Comment by zenparsing** [comment]
Thanks, @pleath I was wondering about that. I'll update with a fix.

File: lib/Runtime/ByteCode/ByteCodeGenerator.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (boingoing) [reply lib/Runtime/ByteCode/ByteCodeGenerator.cpp]: So the point of registering these constants is just to pass the values so we can compare them to some other value in a register or such when executing the bytecode?

**Comment by rhuanjl** [comment]
@zenparsing do you think you'll be able to update this so we can land it? Or should I pull it down, update for the comments above and open a new PR?

**Comment by zenparsing** [comment]
@rhuanjl This PR was already updated per @pleath's comments above. I'll try to rebase, regenerate bytecode, and give it another review within a couple of days.

**Comment by rhuanjl** [comment]
@zenparsing Please could you rebase this and regen the bytecode? I'll give it another review - I think I'm familiar enough now with most of this logic.

(I've got some follow up PRs I want to open)

**Comment by rhuanjl** [comment]
Merged as https://github.com/microsoft/ChakraCore/commit/c848d4d8d50c0dfb4a23540a9ee6cd023fa029c1

**Pull request closed without merging**